object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/CalledMethodsPredicateEvaluator.java

```diff
@@ -15,6 +29,118 @@ public CalledMethodsPredicateEvaluator(final Set<String> cmMethods) {
+        context.getFormulaManager().getBooleanFormulaManager();
+    BooleanFormula lhsBool, rhsBool;
+
+    try {
+      lhsBool = compilationUnitToBooleanFormula(lhsAst, booleanFormulaManager);
+      rhsBool = compilationUnitToBooleanFormula(rhsAst, booleanFormulaManager);
+    } catch (UnsupportedOperationException e) {
+      return false;
+    }
+
+    BooleanFormula satQuery =
+        booleanFormulaManager.not(booleanFormulaManager.implication(lhsBool, rhsBool));
+
+    ProverEnvironment prover =
+        context.newProverEnvironment(SolverContext.ProverOptions.GENERATE_MODELS);
+
+    try {
+      prover.addConstraint(satQuery);
+    } catch (InterruptedException e) {
+      return false;
+    }
+
+    try {
+      return prover.isUnsat();
+    } catch (Exception e) {
+      return false;
+    }
+  }
+
+  /**
+   * Converts a Javaparser AST representing a boolean expression into a boolean formula that could
+   * be passed to an SMT solver.
+   */
+  private static BooleanFormula compilationUnitToBooleanFormula(
+      CompilationUnit ast, BooleanFormulaManager booleanFormulaManager) {
+
+    BlockStmt theBlock = ast.getType(0).getMembers().get(0).asMethodDeclaration().getBody().get();
+
+    com.github.javaparser.ast.expr.Expression theExpression =
+        theBlock.getStatements().get(0).asReturnStmt().getExpression().get();
+
+    try {
+      return expressionToBooleanFormula(theExpression, booleanFormulaManager);
+    } catch (UnsupportedOperationException e) {
+      throw e;
+    }
+  }
+
+  /** Actual implementation of conversion between Javaparser expression and boolean formula. */
+  private static BooleanFormula expressionToBooleanFormula(
+      com.github.javaparser.ast.expr.Expression theExpression,
+      BooleanFormulaManager booleanFormulaManager) {
+    if (theExpression.isNameExpr()) {
+      return booleanFormulaManager.makeVariable(theExpression.asNameExpr().getNameAsString());
+    } else if (theExpression.isBinaryExpr()) {
+      if (theExpression.asBinaryExpr().getOperator().equals(BinaryExpr.Operator.OR)) {
+        return booleanFormulaManager.or(
+            expressionToBooleanFormula(
+                theExpression.asBinaryExpr().getLeft(), booleanFormulaManager),
+            expressionToBooleanFormula(
+                theExpression.asBinaryExpr().getRight(), booleanFormulaManager));
+      } else if (theExpression.asBinaryExpr().getOperator().equals(BinaryExpr.Operator.AND)) {
+        return booleanFormulaManager.and(
+            expressionToBooleanFormula(
+                theExpression.asBinaryExpr().getLeft(), booleanFormulaManager),
+            expressionToBooleanFormula(
+                theExpression.asBinaryExpr().getRight(), booleanFormulaManager));
+      }
+    } else if (theExpression.isEnclosedExpr()) {
+      return expressionToBooleanFormula(
+          theExpression.asEnclosedExpr().getInner(), booleanFormulaManager);
+    }
+    throw new UnsupportedOperationException();
```

Comment:
This should be more informative; at the least, it should pass back either `theExpression` or a string representation of it.